File: solution/2326. Spiral Matrix IV/2326.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        vector<vector<int>> res(m, vector<int>(n, -1));
        int y = 0, x = -1;
        int dy = 0, dx = 1;
        for (; head; head = head->next) {
            int ny = y + dy;
            int nx = x + dx;
            if (ny < 0 || ny >= m || nx < 0 || nx >= n || res[ny][nx] != -1) {
                swap(dy, dx);
                dx = -dx;
            }
            y += dy;
            x += dx;
            res[y][x] = head->val;
        }
        return res;
    }
};

// dy, dx
//  0,  1
//  1,  0
//  0, -1
// -1,  0

// swap(dy, dx)
// dx = -dx

File: solution/2326. Spiral Matrix IV/2326.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        res = [[-1] * n for _ in range(m)]
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        d = 0
        y, x = 0, -1
        while head:
            dy, dx = dirs[d]
            ny = y + dy
            nx = x + dx
            if not (0 <= ny < m and 0 <= nx < n and res[ny][nx] == -1):
                d = (d + 1) % 4
            dy, dx = dirs[d]
            y += dy
            x += dx
            res[y][x] = head.val
            head = head.next
        return res
        
